extension/src/services/AvailableProjectsService.ts

```diff
@@ -0,0 +1,235 @@
+import * as vscode from 'vscode';
+import { extensionLogOutputChannel } from '../utils/logging';
+import path from "path";
+
+// Best effort attempt to provide an available list of runnable projects, including single file AppHosts
+// as calling into the aspire cli to perform this action is too slow, especially on large codebases
+export class AvailableProjectsService implements vscode.Disposable {
+    private _projects: Map<vscode.WorkspaceFolder, Set<string> | null> = new Map();
+    private _fileWatchers: vscode.FileSystemWatcher[] = [];
+    private _folderWatchers: Map<vscode.WorkspaceFolder, vscode.FileSystemWatcher[]> = new Map();
+    private _disposables: vscode.Disposable[] = [];
+
+    constructor() {
+        // Initialize AppHost candidates for all workspace folders
+        this.initializeAppHostCandidates();
+
+        // Set up file watchers for AppHost.cs and *.csproj files
+        this.setupFileWatchers();
+
+        // Watch for workspace folder changes
+        this.setupWorkspaceWatchers();
+    }
+
+    dispose(): void {
+        this._fileWatchers.forEach(watcher => watcher.dispose());
+        this._folderWatchers.forEach(watchers => watchers.forEach(watcher => watcher.dispose()));
+        this._disposables.forEach(disposable => disposable.dispose());
+    }
+
+    public getProjects(folder: vscode.WorkspaceFolder): string[] | null {
+        const projectSet = this._projects.get(folder);
+        return projectSet ? Array.from(projectSet) : null;
+    }
+
+    public async getAppHostCandidates(folder: vscode.WorkspaceFolder): Promise<string[]> {
+        // Return cached candidates if available
+        const cached = this._projects.get(folder);
+        if (cached !== undefined) {
+            return cached ? Array.from(cached) : [];
+        }
+
+        // If not cached, compute and cache them
+        try {
+            const candidates = await this.computeAppHostCandidates(folder);
+            const candidateSet = new Set(candidates);
+            this._projects.set(folder, candidateSet);
+            return candidates;
+        } catch (error) {
+            extensionLogOutputChannel.error(`Error computing app hosts for ${folder.name}: ${error}`);
+            this._projects.set(folder, null);
```

Comment:
Do we get any value out of storing `null` when we couldn't resolve the AppHosts in the workspace? It feels like a sensible structure would be:

- Key exists in Map and is not null/undefined: there's an apphost project here
- Key exists in Map but is null: this workspace has no apphosts (maybe use an empty set here to be more explicit?)
- Key does not exist in Map: there's nothing in the cache